plugins/dist_table_procedure.cc

```diff
@@ -0,0 +1,581 @@
+    : Procedure(owner)
+{
+    procConfig = config.params.convert<DistTableProcedureConfig>();
+}
+
+Any
+DistTableProcedure::
+getStatus() const
+{
+    return Any();
+}
+
+RunOutput
+DistTableProcedure::
+run(const ProcedureRunConfig & run,
+      const std::function<bool (const Json::Value &)> & onProgress) const
+{
+    DistTableProcedureConfig runProcConf =
+        applyRunConfOverProcConf(procConfig, run);
+
+    SqlExpressionMldbScope context(server);
+    auto boundDataset = runProcConf.trainingData.stm->from->bind(context);
+
+    vector<Utf8String> outcome_names;
+    for(const pair<string, std::shared_ptr<SqlExpression>> & lbl : runProcConf.outcomes)
+        outcome_names.push_back(lbl.first);
+
+    DistTablesMap distTablesMap;
+
+    {
+        // Find only those variables used
+        SqlExpressionDatasetScope context(boundDataset);
+
+        auto selectBound = runProcConf.trainingData.stm->select.bind(context);
+
+        for (auto & c: selectBound.info->getKnownColumns()) {
+            distTablesMap.insert(make_pair(c.columnName, DistTable(c.columnName, outcome_names)));
+            cerr << "initializing a stats table for column " 
+                 << c.columnName << endl;
+        }
+    }
+
+    auto onProgress2 = [&] (const Json::Value & progress)
+        {
+            Json::Value value;
+            value["dataset"] = progress;
+            return onProgress(value);
+        };
+
+    auto output = createDataset(server, runProcConf.output, onProgress2, true /*overwrite*/);
+
+    int num_req = 0;
+    Date start = Date::now();
+
+    // columns cache
+    // key: feature column name
+    // values: in order, column name for 
+    map<ColumnName, vector<ColumnName>> colCache;
+
+    const int nbOutcomes = runProcConf.outcomes.size();
+
+    auto processor = [&] (NamedRowValue & row_,
+                           const std::vector<ExpressionValue> & extraVals)
+        {
+            MatrixNamedRow row = row_.flattenDestructive();
+
+            if(num_req++ % 5000 == 0) {
+                double secs = Date::now().secondsSinceEpoch() - start.secondsSinceEpoch();
+                string progress = ML::format("done %d. %0.4f/sec", num_req, num_req / secs);
+                onProgress(progress);
+                cerr << progress << endl;
+            }
+
+            // we parse in advance the value for each outcome
+            vector<bool> haveTarget(nbOutcomes);
+            vector<double> targets(nbOutcomes);
+            for(int i=0; i < nbOutcomes; i++) {
+                CellValue outcome = extraVals.at(i).getAtom();
+                haveTarget[i] = !outcome.empty();
+                if (haveTarget[i])
+                    targets[i] = outcome.toDouble();
+            }
+
+            map<ColumnName, size_t> column_idx;
+            for(int i=0; i<row.columns.size(); i++) {
+                column_idx.insert(make_pair(get<0>(row.columns[i]), i));
+            }
```

Comment:
http://stackoverflow.com/questions/13049340/initializing-a-stdmap-when-the-size-is-known-in-advance
non?